src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs

```diff
@@ -96,21 +110,21 @@ private async Task<Package> PreparePackagesAsync(PackageType requestedPackageTyp
                 var packageTypes = GetPackageTypes(requestedPackageType);
 
                 // Push all of the package types that have not been pushed yet.
-                var pushTasks = new Dictionary<PackageType, Task<Package>>();
+                var pushTasks = new Dictionary<PackageType, Task<List<Package>>>();
                 lock (_packagesLock)
                 {
                     foreach (var packageType in packageTypes)
                     {
-                        if (!_packages.TryGetValue(packageType, out Package package))
+                        if (!_packages.TryGetValue(packageType, out IReadOnlyList<Package> package))
                         {
                             pushTasks.Add(packageType, UncachedPrepareAsync(requestedPackageType, packageType, logger));
                         }
                     }
                 }
 
                 await Task.WhenAll(pushTasks.Values);
-                
-                // Use the package that we just pushed.
+
+                // Use the packages that we just pushed.
                 return _packages[requestedPackageType];
             }
```

Comment:
Shouldn't this line be protected by the `_packagesLock`? Someone else could be mutating `_packages` here

Reply:
This is an existing code path, and I don't think it needs lock, the lock and the logic before ensures that the _packages is writable only once for the corresponding `requestedPackageType`.

Reply:
You cannot read a dictionary as it is being mutated, regardless of the key

Reply:
The logic ensures that the dictionary mutation happens before any thread could possibly read it here. However, it doesn't seem to be protected by any mutation from outside of this method(we don't have any today). I'll add the lock for possible future degradation.